src/driver/safe/external_memory.rs

```diff
@@ -0,0 +1,124 @@
+        }?;
+        #[cfg(windows)]
+        let external_memory = unsafe {
+            use std::os::windows::io::AsRawHandle;
+            result::external_memory::import_external_memory_opaque_win32(file.as_raw_handle(), size)
+        }?;
+        Ok(ExternalMemory {
+            external_memory,
+            size,
+            _device: self.clone(),
+            _file: ManuallyDrop::new(file),
+        })
+    }
+}
+
+/// An abstraction for imported external memory.
+///
+/// This struct can be created via [`CudaDevice::import_external_memory`].
+/// The imported external memory will be destroyed when this struct is dropped.
+#[derive(Debug)]
+pub struct ExternalMemory {
+    external_memory: sys::CUexternalMemory,
+    size: u64,
+    _device: Arc<CudaDevice>,
+    _file: ManuallyDrop<File>,
+}
+
+impl Drop for ExternalMemory {
+    fn drop(&mut self) {
+        unsafe { result::external_memory::destroy_external_memory(self.external_memory) }.unwrap();
+        // From CUDA docs, when successfully importing UNIX file descriptor:
+        // Ownership of the file descriptor is transferred to the CUDA driver when the handle is imported successfully.
+        // Performing any operations on the file descriptor after it is imported results in undefined behavior.
+        #[cfg(not(unix))]
+        unsafe {
+            ManuallyDrop::<File>::drop(&mut self._file)
+        };
+    }
+}
+
+impl ExternalMemory {
+    /// Map the whole external memory to get mapped buffer.
+    pub fn map_all(&mut self) -> Result<MappedBuffer<'_>, DriverError> {
+        self.map_range(0..self.size as usize)
+    }
+
+    /// Map a range of the external memory to a mapped buffer.
+    ///
+    /// Only one mapped buffer is allowed at a time.
+    /// This is more restrictive than it necessarily needs to be,
+    /// but it makes enforcing safety easier.
+    ///
+    /// # Panics
+    /// This function will panic if the range is invalid,
+    /// such as when the start or end is larger than the size.
+    pub fn map_range(&mut self, range: Range<usize>) -> Result<MappedBuffer<'_>, DriverError> {
+        assert!(range.start as u64 <= self.size);
+        assert!(range.end as u64 <= self.size);
+        let device_ptr = unsafe {
+            result::external_memory::get_mapped_buffer(
+                self.external_memory,
+                range.start as u64,
+                range.len() as u64,
+            )
+        }?;
+        Ok(MappedBuffer {
+            device_ptr,
+            len: range.len(),
+            _marker: PhantomData,
+        })
+    }
+}
+
+/// An abstraction for a mapped buffer for some external memory.
+///
+/// This struct can be created via [`ExternalMemory::map_range`] or [`ExternalMemory::map_all`].
+/// The underlying mapped buffer will be freed when this struct is dropped.
+#[derive(Debug)]
+pub struct MappedBuffer<'a> {
+    device_ptr: sys::CUdeviceptr,
+    len: usize,
+    _marker: PhantomData<&'a mut ExternalMemory>,
```

Comment:
Thoughts on actually having a reference here? If we do this then in the drop implementation, we could use `result::free_async(self.device_ptr, self.external_memory.device.stream)`

Reply:
I am unsure if that is sound, because from the [CUDA docs](https://docs.nvidia.com/cuda/cuda-driver-api/group__CUDA__EXTRES__INTEROP.html#group__CUDA__EXTRES__INTEROP_1gb9fec33920400c70961b4e33d838da91) it explicitly mentions that `cuMemFree` must be used:

> The returned pointer devPtr must be freed using [cuMemFree](https://docs.nvidia.com/cuda/cuda-driver-api/group__CUDA__MEM.html#group__CUDA__MEM_1g89b3f154e17cc89b6eea277dbdf5c93a).

Reply:
Oh great point, can you add a note in the code about that?

Reply:
I added a note to `get_mapped_buffer`.